Devrun script (uses iojs)

ochafik opened the pull request:
Note that this requires / enforces a [next-nightly iojs build](https://iojs.org/download/next-nightly/). The version of V8 that currently ships with it ([4.4.63](https://chromium.googlesource.com/v8/v8.git/+/refs/heads/4.4.63).12 as of iojs-v3.0.0-next-nightly20150722cf9e02daf7, can be checked with `iojs -e 'console.log(process.versions.v8);'`) still lacks some major ES6 features (lexical this in arrow-functions, for instance), but simple "hello world"-style program already run fine.

jmesserly commented:
Awesome!!! Going to take a look.

Would you mind squashing these to a single commit?
(if I'm doing a Pull Request I usually do `git rebase -i origin/master' and then change change "pick" to "s" for squash. alternatively [depot_tools](https://www.chromium.org/developers/how-tos/install-depot-tools) adds the commands `git cl upload` and `git cl land` for working with https://codereview.chromium.org, the latter command will squash automatically when it checks in)

jmesserly commented on lib/runtime/dart/_interceptors.js:
```diff
@@ -391,7 +391,7 @@ dart_library.library('dart/_interceptors', null, /* Imports */[
       }
     }
-    dart.setBaseClass(JSArray, dart.global.Array);
+    dart.setBaseClass(JSArray, (typeof global == 'undefined' ? dart.global : global).Array);
```
the idea behind `dart.global` was to be the global object (`window` in browser or `global` on node), was that not working for you? I'd like to find a better solution than expanding this in place. I bet we can do it though.

ochafik replied:
Actually, it all works without it, not sure what led to this TBH. I was having difficulties aliasing global to dart.global (dart.global.Array would always return undefined). Might have been fixed by the `let window = window;` fix.

jmesserly commented on lib/runtime/dart_library.js:
```diff
@@ -5,7 +5,7 @@
 */
 
-var dart_library;
+var dart_library = typeof module !== "undefined" ? module.exports : undefined;
```
can we find a module pattern that works for browser and io.js?
(this here seems like it could be folded into the code below: `dart_library || (dart_library = {})` at the least)

ochafik replied:
I've folded things the other way around for now: `var dart_library = typeof module != "undefined" && module.exports || {};`, keeping nicer module patterns as a snack for later if that's okay :-)

jmesserly commented on lib/src/compiler.dart:
```diff
@@ -307,6 +307,8 @@ abstract class AbstractCompiler {
       filepath = 'dart/$filepath';
     } else if (uri.scheme == 'file') {
       filepath = path.relative(filepath, from: inputBaseDir);
+      // Play nice with paths not under inputBaseDir (which may be ''):
```
can you explain the thinking here?
do we need to offer inputBaseDir as an option?

ochafik replied:
This was a dirty hack needed when the input file was a relative file. Fixed by absolutizing the path in AbstractCompiler.stringToUri

jmesserly commented on bin/devc.dart:
```diff
@@ -20,31 +21,40 @@ void _showUsageAndExit() {
 }
 
-void main(List<String> args) {
+CompilerOptions validateOptions(List<String> args) {
```
is there a functional change to this file, or just restructuring?

jmesserly commented on bin/devc.dart:
```diff
@@ -20,31 +21,40 @@ void _showUsageAndExit() {
+  return options;
+}
 
+Future<bool> compile(CompilerOptions options) async {
```
nit: `async` not needed if you don't use `await`

ochafik replied:
In this case there's a hidden conversion, BatchCompiler.run()'s bool return is promoted to a future thanks to this async: added a line of comment (moved to `lib/src/compiler.dart`)

jmesserly commented on bin/devrun.dart:
```diff
@@ -0,0 +1,104 @@
+    _showUsageAndExit();
+  }
+  if (options.codegenOptions.outputDir == null) {
+    // If the user didn't specify an output directory, create a temp one.
```
we could just make the temp dir always?

ochafik replied:
I'd rather move this to parseOptions, passed a forceOutDir: true through to there.

jmesserly commented on bin/devrun.dart:
```diff
@@ -0,0 +1,104 @@
+
+import 'package:path/path.dart';
+
+import 'devc.dart' show validateOptions, compile;
```
interesting. don't usually see importing another bin/ script. But I can't think of anything wrong with it, so maybe it's a valid pattern. it can be hard(er) to test "bin/" though, so we usually keep them really small, with most of the logic someplace else (e.g. lib/src/compiler.dart)

ochafik replied:
Moved the common utils to `lib/src/compiler.dart` :-)

jmesserly commented on lib/src/runner/runtime_utils.dart:
```diff
@@ -0,0 +1,80 @@
+  });
+}
+
+List<String> _turnFilesToUris(List<String> args) =>
```
this is a copy of turnFilesToUris from another place?

jmesserly commented on lib/src/server/server.dart:
```diff
@@ -257,7 +257,8 @@ class DevServer {
     await shelf.serve(handler, host, port);
     print('Serving $_entryPath at http://$host:$port/');
-    compiler.run();
+    CheckerResults results = compiler.run();
```
nice catch. I think this was supposed to return Future<CheckerResults> but I think CheckerResults is deprecated at this point, so +1 just returning status.

jmesserly commented on tool/input_sdk/private/isolate_helper.dart:
```diff
@@ -697,12 +697,14 @@ class _IsolateEvent {
     // and: http://www.w3.org/TR/Window/#dfn-self-attribute
-    JS("void", r"self.postMessage(#)", msg);
+    JS("void",
+       r"(typeof global == 'undefined' ? self : global).postMessage(#)", msg);
```
should probably save this into a variable and use it.

final _global = JS('...');

elsewhere:

    return JS('', '#.postMessage(#)', _global, msg);

if we can find a way to fix `dart.global` that makes it even easier:

    return JS('', 'dart.global.postMessage(#)', _global, msg);

ochafik replied:
Sweet, done!

jmesserly commented on bin/devrun.dart:
```diff
@@ -0,0 +1,104 @@
+
+main(List<String> args) async {
+  // TODO(ochafik): Should devc accept files anyway?
+  args = turnFilesToUris(args);
```
wouldn't we want to do this after we parse options? Otherwise we make system calls to discover "--help" is not really a file. And if it was, then --help would stop working.

ochafik replied:
Dropped hack in favour of AbstractCompiler.stringToUri tweak

jmesserly commented on bin/devrun.dart:
```diff
@@ -0,0 +1,104 @@
+}
+
+main(List<String> args) async {
+  // TODO(ochafik): Should devc accept files anyway?
```
I don't understand the comment?

ochafik replied:
devc didn't seem to accept relative files (due to the way the URI is built in AbstractCompiler.stringToUri)

jmesserly commented on lib/src/runner/file_utils.dart:
```diff
@@ -0,0 +1,23 @@
+  return list;
+}
+
+List<String> turnFilesToUris(List<String> args) =>
```
what's the reason to do this? we should have logic already to handle relative path -> file URI as needed
(stringToUri in compiler.dart will do this). I've been desperately trying to tame all of the path logic that sprung up in our code base, so apologies in advance for focusing so much on this

jmesserly replied:
(and maybe the answer is to handle it immediately when we parse options, then deal with URIs from there on out, but should only do the operation on files.)

ochafik replied:
hehe, apologies for not doing more thorough checks for alternative fixes before sending this PR ;-)

jmesserly commented on lib/src/runner/file_utils.dart:
```diff
@@ -0,0 +1,23 @@
+    args.map((arg) {
+      var file = new File(arg);
+      return file.existsSync() ? "file:" + file.absolute.path : arg;
+    }).toList();
```
missing newline?

jmesserly commented on lib/src/runner/iojs_utils.dart:
```diff
@@ -0,0 +1,55 @@
+        "Please visit https://iojs.org/download/next-nightly/");
+    exit(1);
+  }
+  stderr.writeln("WARNING: io.js does not fully support some critical "
```
feel free to restore the `=>` `this` workaround if you'd like:
https://github.com/dart-lang/dev_compiler/commit/19f6daf59b49e425961b01e34ebd50339f8f90a1#diff-3a4f8912d1b55456831b2e7b40ba8d68L847

should be pretty easy to pass an option down, so Printer only inserts the workaround if needed.
it was fixed a while ago in V8/Chrome, though, so maybe iojs will get it soon

jmesserly replied:
oh, btw, I'm happy to help restoring that workaround too. Let me know!

ochafik replied:
This would help a lot, TBH :-)

jmesserly commented on lib/src/runner/iojs_version.dart:
```diff
@@ -0,0 +1,59 @@
+///
+/// Note that io.js is the future of node.js:
+/// http://www.infoq.com/news/2015/05/nodejs-iojs
+class IoJsVersion extends Comparable {
```
FYI, pub_semver package has a Version object. We use it in our SDK version check script:
https://github.com/dart-lang/dev_compiler/blob/902cab3c54226b28548140ada0c59d4da48cb8c2/tool/sdk_version_check.dart#L21

ochafik replied:
Awesome, works great!

jmesserly commented on lib/src/runner/runtime_utils.dart:
```diff
@@ -0,0 +1,80 @@
+/// In node.js / io.js, these modules need to be aliased globally
+/// (e.g. `var foo = require('./path/to/foo.js')`).
+const _ALIASED_RUNTIME_FILES = const {
+  'dart_library.js': 'dart_library',
```
another option here, we could change our module pattern, to make these imports explicit if we're in nodejs mode.
(in ES6 we'll want to do that too. The current DDC module pattern is a bit of a hack...)

ochafik replied:
Left a TODO for now, until I understand ES6 modules better :-)

jmesserly commented on lib/src/runner/runtime_utils.dart:
```diff
@@ -0,0 +1,80 @@
+
+String getMainModuleName(CompilerOptions options) {
+  // This is an ugly hack to guess the path to the output entry point .js.
+  // TODO(ochafik): Find / create a clean way to get the output .js file(s).
```
I think you want getModulePath and/or getOutputPath
https://github.com/dart-lang/dev_compiler/blob/0b0f07faea680f4f8b3ffaa8854f82be4febc960/lib/src/compiler.dart#L309

ochafik replied:
Indeed, although I'm unable to find a beautiful way to split them out for reuse right now (I don't have a handle on an AbstractCompiler yet). Simplified the logic a lot anyway (with new absolute file urls) and left a todo.

jmesserly commented:
Added some comments but this is so cool :+1:

ochafik commented:
Squashed to a single commit, PTAL :-)

jmesserly commented on lib/src/compiler.dart:
```diff
@@ -36,6 +36,7 @@ import 'info.dart'
     show AnalyzerMessage, CheckerResults, LibraryInfo, LibraryUnit;
 import 'options.dart';
 import 'report.dart';
+import 'package:dev_compiler/src/server/server.dart';
```
trivial, but this should probably be "src/server/server.dart" to match the style of the other imports

jmesserly replied:
I'd also be tempted to "show DevServer" to make it clear that's the only part used

ochafik replied:
Done

jmesserly commented on lib/src/compiler.dart:
```diff
@@ -46,6 +47,37 @@ StreamSubscription setupLogger(Level level, printFn) {
+    var context = createAnalysisContextWithSources(
+        options.strongOptions, options.sourceOptions);
+    var reporter = createErrorReporter(context, options);
+    // Note: run returns a bool, turned into a future since this function is async.
```
yeah, fwiw, not sure comment is needed, I think I was just confused.

jmesserly replied:
(fine to leave it there though)

jmesserly commented:
looks like it needs to be rebased against latest, otherwise LGTM

ochafik commented:
Rebased against latest (now using defaultRuntimeFiles, added '_generators.js' to it): runs DeltaBlue out of the box (thanks to John's arrow-fun-binding workaround reinstatement).

ochafik commented:
Sorry, did one last refactoring: realized it was easy enough to support d8 as well. Changed the version check to pull out the version of v8 instead of that of iojs (`d8 '-e' 'print(version())'` or `iojs '-p' 'process.versions.v8'`), and added a `--v8-binary` option (under new `RunnerOptions`).

jmesserly commented:
LGTM :+1:
